Vastly improve shroud rendering performance.

RoosterDragon opened the pull request:
Changes in the shroud are now tracked. If a cell changes it will mark itself and its neighbours as dirty. During the render phase all dirty cells will have their vertices calculated and cached. If a cell is not dirty, the pre-calculated vertices are retrieved from cache. Then the sprite renderer is provided the sprite and the pre-calculated vertices to draw.

This prevents constant recalculation of vertices for the shroud in the render phase, requiring instead only dirty cells in the visible area. The update phase is reduced to a practical noop, instead incurring the cost only of changed cells each frame, rather than checking the visible area.

---

This is a revival of #7095 with some updates. I initially thought @pchote's proof on concept would be the same for performance, and it was far less code so I thought it'd be a better choice.

Sadly it's not. You need to update the cell and its 8 neighbours. If you generate vertices right away you end up doing a substantial amount of work that can easily be avoided by delaying the work until you actually need to render the vertices.

I tested this version, and saw total CPU time for rendering the shroud drop from 28% (told you it was expensive!) to 9% of all CPU time spent.

In theory, this means if you get a solid 50 fps, you could now be getting 60 fps instead!

RoosterDragon committed:
Vastly improve shroud rendering performance.

Changes in the shroud are now tracked. If a cell changes it will mark itself and its neighbors as dirty. During the render phase all dirty cells will have their vertices calculated and cached. If a cell is not dirty, the pre-calculated vertices are retrieved from cache. Then the sprite renderer is provided the sprite and the pre-calculated vertices to draw.

This prevents constant recalculation of vertices for the shroud in the render phase, requiring instead only dirty cells in the visible area. The update phase is reduced to a practical noop, instead incurring the cost only of changed cells each frame, rather than checking the visible area.

pchote commented:
I don't want to give up on the vertex buffer approach yet - we can avoid repeating work by keeping a list of dirty cells, and then updating the buffer in one go (e.g. during the next render tick).  We can do further optimizations to that approach, such as tiling and batch-updating the vertex buffer, and all of that can be pulled into a reusable class that can be used by the terrain, shroud, and other terrain layers.

RoosterDragon commented:
I'd much rather keep this PR as a cohesive unit dealing with shroud changes, and have another PR that dealt with some vertex buffer infrastructure as and when we needed it - if that's possible.

huwpascoe commented:
How about using a PBO "[Pixel Buffer Object](http://www.songho.ca/opengl/gl_pbo.html)" instead?

Every tilemap index is stored as a 16bit pixel (or 8bit if the atlas has < 255 tiles) and rendered by a shader onto a single quad.

pchote commented:
Pixel buffers are part of modern GL, and may not be implemented on all of our currently supported hardware.  It's would also add parallel code to solve a problem which we are already solving (for terrain) using VBO's.

@RoosterDragon: My main objection is that the bulk of the changes are are in creating and managing a software-emulated vertex buffer.  This adds even more complexity to what's already a fiddly and fragile system :(

chrisforbes commented:
For hardware support -- this shouldn't be a problem. PBO came in with GL
2.1, which we already need for other things.

On Sun, Jan 11, 2015 at 11:40 AM, Paul Chote <notifications@github.com>
wrote:

> Pixel buffers are part of modern GL, and may not be implemented on all of
> our currently supported hardware. It's would also add parallel code to
> solve a problem which we are already solving (for terrain) using VBO's.
>
> @RoosterDragon <https://github.com/RoosterDragon>: My main objection is
> that the bulk of the changes are are in creating and managing a
> software-emulated vertex buffer. This adds even more complexity to what's
> already a fiddly and fragile system :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/7305#issuecomment-69475123>.
>

huwpascoe commented:
@pchote well the idea is you no longer use any vertices and let the shader do the work. So it's free occlusion culling on top of not having to deal with redundant float[5] xyzuv coordinates unless I'm missing something?

chrisforbes commented:
How exactly do you want to use PBO? All it gives you is buffer object
sourcing for TexImage/TexSubImage/GetTexImage.

On Sun, Jan 11, 2015 at 6:18 PM, yesfish <notifications@github.com> wrote:

> @pchote <https://github.com/pchote> well the idea is you no longer use
> any vertices and let the shader do the work. So it's free occlusion culling
> on top of not having to deal with redundant float[5] xyzuv coordinates
> unless I'm missing something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/7305#issuecomment-69484106>.
>

huwpascoe commented:
It's pretty much the same as using any other blending map but with colors referring to an index rather than a gradient, ranging from 0 - texelformat.maxvalue

Here's an exaggerated example
![Imgur](http://i.imgur.com/4L9B3Yo.jpg)

RoosterDragon commented:
I'm a bit lost on how best to do this directly into a vertex buffer without just creating more problems for myself.

The renderer already solves the problem of setting the correct texture and blend state - but if I want to keep a vertex buffer I either need to track those states myself and duplicate the effort of ensuring the correct render state, or I somehow need to ensure those states are invariant over all the shroud and fog sprites.

Is there a method of doing this that doesn't involve a bunch of extra work?

Also, it seems like to render this shroud we'll need to issue draw calls for each row (so we can avoid drawing the offscreen extents to the left and right) or else accept rendering too much (just render all visible rows, including the parts that extend offscreen, but this at least allows us to render a single contiguous chunk of buffer).

Need some help here.

chrisforbes commented:
To address the last point -- Instead of doing things row-granular we'd be
much better to add some tiling -- 8x8 or 16x16 are probably good starting
points. We can also look at MultiDrawArrays / MultiDrawElements to reduce
validation overhead a little, where we're forced to draw many batches.

On Wed, Jan 14, 2015 at 8:36 AM, RoosterDragon <notifications@github.com>
wrote:

> I'm a bit lost on how best to do this directly into a vertex buffer
> without just creating more problems for myself.
>
> The renderer already solves the problem of setting the correct texture and
> blend state - but if I want to keep a vertex buffer I either need to track
> those states myself and duplicate the effort of ensuring the correct render
> state, or I somehow need to ensure those states are invariant over all the
> shroud and fog sprites.
>
> Is there a method of doing this that doesn't involve a bunch of extra work?
>
> Also, it seems like to render this shroud we'll need to issue draw calls
> for each row (so we can avoid drawing the offscreen extents to the left and
> right) or else accept rendering too much (just render all visible rows,
> including the parts that extend offscreen, but this at least allows us to
> render a single contiguous chunk of buffer).
>
> Need some help here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/7305#issuecomment-69804929>.
>

pchote commented:
Using a proper vertex buffer does indeed restrict you to having all the sprites on the same sheet, and this is something that needs to be managed manually -- regular sequences won't guarantee this.

The easiest solution would be to move the shroud definition into the tileset, which does guarantee a single sheet.

The terrain renderer issues a single draw call that renders all of the rows that are visible on screen.  This does mean that there is a lot of horizontal overdraw, but all the data is already on the GPU so this is probably not a huge deal (but @chrisforbes may prove me wrong).

pchote commented:
If jumping straight to a vbo is going to be too much trouble, then I can live with the current implementation (assuming it is bug free).  We can work on building a simpler framework for buffered terrain layers across future prs.

chrisforbes commented:
@pchote the extra junk we draw off the sides isn't *usually* a huge problem, but it gets worse as the maps get wider.

obrakmann commented:
I didn't spot any regressions in-game, and I won't even try to understand what happens in this code, but I noticed that while this applies cleanly on top of bleed, the compilation still fails, so I guess it needs a rebase:

```
CSC mods/common/OpenRA.Mods.Common.dll
OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs(361,22): error CS1061: Type `OpenRA.Graphics.PaletteReference' does not contain a definition for `Index' and no extension method `Index' of type `OpenRA.Graphics.PaletteReference' could be found. Are you missing an assembly reference?
```

RoosterDragon commented:
Rebased.

pchote commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -108,6 +109,21 @@ public void DrawSprite(Sprite s, float2 a, float2 b, float2 c, float2 d)
+
+			currentBlend = s.BlendMode;
+			currentSheet = s.Sheet;
+			Array.Copy(sourceVertices, offset, vertices, nv, 4);
```
This is essentially inlining the equivalent to the `Util.FastCreateQuad` above.  It would be nice if we could merge or otherwise improve the consistency.

RoosterDragon replied:
True. I've extracted a method for ensuring the render state is set.

pchote commented on OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs:
```diff
@@ -197,108 +215,162 @@ Edges GetObserverEdges(CPos p)
+			// - If the cells remain off-screen for a long time, they may change several times before we next view
+			// them, so calculating their new vertices is wasted effort since we may recalculate them again before we
+			// even get a chance to render them.
+			// - Cells tend to be invalidated in groups (image as a unit moves, it advances a wave of sight and leaves
```
image &rarr; imagine?

RoosterDragon replied:
Yup.

Mailaender commented:
Checked for shroud and fog rendering glitches on all 4 mods. Could not spot any regressions. :white_check_mark:

penev92 commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -49,6 +50,19 @@ public void Flush()
+
+			if (s.Sheet != currentSheet ||
+				s.BlendMode != currentBlend ||
+				nv + 4 > renderer.TempBufferSize)
```
These conditions could easily fit on one line and be easier to read.

RoosterDragon replied:
Yup.

penev92 commented on OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs:
```diff
@@ -65,38 +65,35 @@ enum Edges : byte
+		readonly Vertex[] fogVertices, shroudVertices;
+		readonly Sprite[] fogSprites, shroudSprites;
+		readonly CellLayer<Sprite> fogSpriteLayer, shroudSpriteLayer;
 		PaletteReference fogPalette, shroudPalette;
```
Group this with the other not-readonly variables below.

RoosterDragon replied:
I'm grouping this here because they're semantically related. Everything here is a fog/shroud combo. I'd rather do this that group by readonly-ness unless there's a particular need.

penev92 replied:
Eh, just conveying what I've been told on my PRs.

pchote replied:
Grouping by semantics is more important than readonly/not.

penev92 commented on OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs:
```diff
@@ -197,108 +215,162 @@ Edges GetObserverEdges(CPos p)
+
+		void RenderObserverShroud(CellRegion.MapCoordsRegion renderRegion)
+		{
+			// The observer shroud is always fully cached, so we can just render straight from the cache.
```
What does "fully cached" mean?

RoosterDragon replied:
My intention was to convey the fact that the observer shroud is fixed. We generate all the tiles (i.e. cache) it right away (see UpdateNullShroud above) and so when we need to render it, we can do so without having to check for dirty tiles.

Would something more akin to "The observer shroud never changes and we have cached all the tiles ahead of time - therefore we can render all the cached tiles without checking for any dirty tiles" be clearer?

penev92 replied:
No, not really. What is "observer shroud" then?
By "fully cached" I understand "every single cell on the map is cached".

RoosterDragon replied:
Ah - observer shroud is the shroud you see as an observer when you choose the "Disable shroud" option from the drop-down. Only the edges are shrouded.

Mailaender replied:
Maybe call it border shroud then. The spectator shroud reminds people of the replay shroud switcher which is actually pretty dynamic.

penev92 replied:
Oooh, OK.

penev92 commented:
Due to the low number of people able to review such PRs and their time limitations sometimes less experienced people have to do reviews of stuff they are not familiar with.
Because of that (and for general code readability) I'd ask for method descriptions.
Preferably like this:
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Map/CellRegion.cs#L50
Best like this:
>         /// <summary>
        /// This method does something awesome.
        /// </summary>
        /// <param name="region">A CellRegion to be checked.</param>
        /// <returns>Some description of the result/return value.</returns>
		public bool Contains(CellRegion region)

I think the `ShroudRenderer` could really use some more documenting.

penev92 commented:
Tested all mods \ { TD } both by playing and watching replays. Didn't see any problems.
And nothing struck me as apparently wrong in the code, so :+1:  with the above comments.

Phrohdoh commented:
I don't agree with adding obvious comments such as the example above.

RoosterDragon commented:
I'm somewhat in agreement with Phrohdoh about adding xmldoc on all the methods. I've tried to use short well named methods for readability, and I'm not sure they would add much value since they're generally private to the renderer. (That said, I would personally love more xmldoc in general, but I think some people are worried about redundant commentary)

If the concept here is unclear, I should and certainly can add more descriptive comments about why the shroud renderer has things set up as I have here.

penev92 commented:
Some people that are worried redundant commentary can just not read them.
But I suppose you are right about xmldoc on private methods. I still think some simple comments at least would be good for readability.
If the code changes in this PR were so obvious why wasn't it reviewed a month ago?

pchote commented:
The official stance (in the [coding standard](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard)) is that xml &lt;summary&gt;s are fine where they add extra context, but that parameter explanations should be skipped (because the code should be factored into chunks with self-explanatory arguments).

penev92 commented:
OK, just summaries then.

RoosterDragon commented:
I've added an explanatory comment in RenderPlayerShroud that should hopefully provide some context to future readers.

I've renamed references to the "observer" or "null" shroud to "map border" shroud to hopefully better hint that it is the shroud that affects only the map border.

Hopefully this improves clarity.

penev92 commented:
Thank you :)
`mapBorderShroudIsCached` is much clearer than `clearedForNullShroud`
The commends are not what I expected, but I think they'll do ;)
I'll look at it again tomorrow.

obrakmann commented:
Noticed one problem after all:
![screenshot from 2015-02-20 00 37 03](https://cloud.githubusercontent.com/assets/4331210/6278642/e351e8d2-b898-11e4-927e-724697cfaadd.png)

Only happens when you use the disable shroud cheat on isometric maps. Does not happen when you disable shroud&fog in the lobby, and not on regular maps. The artifacts also don't necessarily appear along the whole length of the border, and their position seems to depend on the spawn point of the player(?).

RoosterDragon commented:
Excellent find.

The regression here is pretty subtle. It was caused when I removed some seemingly redundant code in the WorldLoaded method. That code is actually redundant in most cases but was accidentally fixing a bug in the old GetObserverEdges method.

The bug in that method was basically that it didn't do the same initial "is this cell visible" check as the GetEdges method. This meant tiles outside the map got the wrong shroud.

Fixing this handily makes life even better - we just get rid of the code that treats the observer edges as a special case. That rids us of the bug and means less code to deal with.

obrakmann commented:
Still didn't notice any regressions except the one that has just been fixed, and the code looks understandable even to me, so :+1:

obrakmann commented:
Thanks!  I feel kinda bad that almost all your contributions get lumped into that single entry in the changelog.

Anyway, [Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/fcfb2f235fbba429af10cd213a24ad2563a40573%5E...fcfb2f235fbba429af10cd213a24ad2563a40573)

Mailaender commented:
`Improved shroud rendering performance` should be non-technical enough to be understood by players in a changelog.
